addOns/scripts/src/main/java/org/zaproxy/zap/extension/scripts/ConsolePanel.java

```diff
@@ -359,8 +359,9 @@ private KeyListener getKeyListener() {
                     new KeyListener() {
                         @Override
                         public void keyTyped(KeyEvent e) {
-                            if (script != null && !script.isChanged()) {
-                                extension.getExtScript().setChanged(script, true);
+                            boolean canUndo = getCommandPanel().canUndo();
+                            if (script != null && script.isChanged() != canUndo) {
```

Comment:
Can it ever happen that the script has been changed but the undo history is empty?
This logic will fail in that case.

Reply:
It can, other script properties make the script changed (there's also the case of new scripts). Relevant issue zaproxy/zaproxy#2189.